### PR TITLE
Update ArrayPool docs

### LINKED ISSUE #5813: System.Buffers.ArrayPool.Return has a potentially misleading doc comment
System.Buffers.ArrayPool.Return says this about it's clearArray parameter:

```
If true, the buffer will be cleared of any data; otherwise, the buffer will be returned with any data intact.
```

This is not incorrect, but it's slightly misleading as it implies it'll clear the array of data and then return it if you set this to true. But arrays that are not returned (because they are bigger than the internal buckets account for) are not cleared. Like I said this is understandable (since you aren't supposed to use the array after you give it back, and GC will collect it eventually) but just slightly misleading. I wouldn't suggest changing the code (it'd be a performance hit for sure) but the distinction may matter for very security sensitive code. Using reflection it's possible for someone else to use the array even if you don't (one example would be if it was a field) and passing `true` here may give you the assumption that even in the case of bad behaving code they won't see anything useful there.

It may be appropriate to mention that it will be cleared of data **before anyone else rents it** which not only highlights the potential for it to be not cleared, as well it opens it up so alternate implementations can lazily clear it if they wanted to.

Also the second half of the comment could also be slightly confusing as it's talking from the perspective of the implementation where it uses buckets under the hood and so it's simply delegating the return. The caller sees this method as the "returning" and they are the one who returns it with data intact, passing a parameter to clear it or not.  Perhaps just saying "otherwise, any data in the buffer will remain intact"

### PR DESCRIPTION
**Pull request by stephentoub**
Fix up the XML comments a bit, both in substance and in syntactic correctness.  Also enable them to be built so that they're verified by the compiler and so that a documentation file is generated.

Fixes #5813
cc: @sokket, @mirhagk

### CODE REVIEW

**Comment by mirhagk** [comment]
Much clearer about what's happening now, and nice to see the rest of them cleaned up as well. Thanks for this

**Comment by sokket** [comment]
Great comments :) much better than my original ones. Thanks for updating these! LGTM

**Comment by stephentoub** [comment]
Thanks, Jon and Nathan.

File: src/System.Buffers/src/System/Buffers/ArrayPool.cs [hunk 1]
```diff
@@ -9,24 +9,36 @@
...
+    /// </para>
+    /// <para>
+    /// This class is thread-safe.  All members may be used by multiple threads concurrently.
+    /// </para>
+    /// </remarks>
     public abstract class ArrayPool<T>
     {
         /// <summary>The default number of arrays that are available for rent.</summary>
         private const int DefaultNumberOfArraysPerBucket = 50;
         /// <summary>The default length of each Rent'able array; equal to 1MB in bytes</summary>
```
Comment (benaadams) [inline src/System.Buffers/src/System/Buffers/ArrayPool.cs]: This should be more like `the maximum length of an array that will be eligible for being returned to the pool`. Keeps causing confusion as people think all the arrays given out are 1MB.

Comment (benaadams) [inline src/System.Buffers/src/System/Buffers/ArrayPool.cs]: @stephentoub ping re description change

I misinterpreted here https://github.com/aspnet/benchmarks/pull/44#issuecomment-177759093
And it was again misinterpreted here https://github.com/aspnet/HttpAbstractions/issues/545#issuecomment-179053875

**Pull request merged**